### PR TITLE
Fix GitHub Action Tests

### LINKED ISSUE #1040: Accumulation Table Incompatible with Coverage Tests
I have discovered an error with the accumulation table that occurs when running `pytest --cov python_ta --cov-config=.coveragerc tests` but does not occur when running `pytest tests`. 

In other words, when running pytests with coverage, an error appears related to the accumulation table that does not appear when running the tests normally. 

### Result of running `pytest tests`:

image

### Result of running `pytest --cov python_ta --cov-config=.coveragerc tests`:

final result of running tests with coverage (shows one failure)
test_one_accumulator failure (where it was called)
test_one_accumulator failure (where the IndexError actually occurred)

### PR DESCRIPTION
**Pull request by quiz3**
## Proposed Changes

_(Describe your changes here. Also describe the motivation for your changes: what problem do they solve, or how do they improve the application or codebase? If this pull request fixes an open issue, [use a keyword to link this pull request to the issue](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword).)_

The GitHub Action tests are flaky and often fail on `tests/test_debug/test_accumulation_table.py`. This fixes #1040, regarding the Accumulation Table. The `sys.settrace` calls within `accumulation_table.py` interfere with `sys.settrace` in `pytest`'s coverage testing code.

<details>
<summary>Screenshots of your changes (if applicable)</summary>

</details>

## Type of Change

_(Write an `X` or a brief description next to the type or types that best describe your changes.)_

| Type                                                                                    | Applies? |
| --------------------------------------------------------------------------------------- | -------- |
| 🚨 _Breaking change_ (fix or feature that would cause existing functionality to change) |          |
| ✨ _New feature_ (non-breaking change that adds functionality)                          |          |
| 🐛 _Bug fix_ (non-breaking change that fixes an issue)                                  |          |
| ♻️ _Refactoring_ (internal change to codebase, without changing functionality)          |          |
| 🚦 _Test update_ (change that _only_ adds or modifies tests)                            |          |
| 📚 _Documentation update_ (change that _only_ updates documentation)                    |          |
| 📦 _Dependency update_ (change that updates a dependency)                               |          |
| 🔧 _Internal_ (change that _only_ affects developers or continuous integration)         |     X     |

## Checklist

_(Complete each of the following items for your pull request. Indicate that you have completed an item by changing the `[ ]` into a `[x]` in the raw text, or by clicking on the checkbox in the rendered description on GitHub.)_

Before opening your pull request:

- [x] I have performed a self-review of my changes.
  - Check that all changed files included in this pull request are intentional changes.
  - Check that all changes are relevant to the purpose of this pull request, as described above.
- [x] I have added tests for my changes, if applicable.
  - This is **required** for all bug fixes and new features.
- [x] I have updated the project documentation, if applicable.
  - This is **required** for new features.
- [x] I have updated the project Changelog (this is required for all changes).
- [x] If this is my first contribution, I have added myself to the list of contributors.

After opening your pull request:

- [x] I have verified that the pre-commit.ci checks have passed.
- [x] I have verified that the CI tests have passed.
- [ ] I have reviewed the test coverage changes reported by Coveralls.
- [x] I have [requested a review](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/proposing-changes-to-your-work-with-pull-requests/requesting-a-pull-request-review) from a project maintainer.

## Questions and Comments

_(Include any questions or comments you have regarding your changes.)_

Were you expecting the tests to fail? As far as I understood the assignment, the tests should now work correctly.

### CODE REVIEW

**Comment by quiz3** [comment]
I believe I have done what you asked in Slack. Please correct me if I have not understood correctly.

**Comment by quiz3** [comment]
> I believe I have done what you asked in Slack. Please correct me if I have not understood correctly.

Utku just clarified with me that I did _not_ correctly interpret your instructions. I will go back and re-do this PR.

**Comment by quiz3** [comment]
NOTE: I think setting draft PRs to run tests is what I was asked to do, but it's possible I have misunderstood.

File: .github/workflows/test.yml [hunk 1]
```diff
@@ -9,7 +9,7 @@ on:
 
 jobs:
   test:
-    if: github.event.pull_request.draft == false
+    # if: github.event.pull_request.draft == false
```
Comment (david-yz-liu) [inline .github/workflows/test.yml:12]: Yes the correct change is to move this line, but it should be done in a separate pull request since it's unrelated to the change you're working on.

File: .github/workflows/test.yml [hunk 2]
```diff
@@ -38,8 +38,8 @@ jobs:
           pip install -e .[dev,cfg,z3]
       - name: Run tests
         run: |
-          pytest --cov python_ta --cov-config=.coveragerc --cov-report lcov --ignore=tests/test_type_constraints --ignore=tests/test_type_inference tests --ignore=tests/test_debug
-          pytest tests/test_debug
+          pytest tests --cov python_ta --cov-config=.coveragerc --cov-report lcov --ignore=tests/test_type_constraints --ignore=tests/test_type_inference --ignore=tests/test_debug
```
Comment (david-yz-liu) [inline .github/workflows/test.yml:41]: This isn't the correct change. You don't need to move `tests`, but you _do_ need to modify the `--ignore=tests/test_debug` part. (Re-read my fourth bullet point carefully.)

Comment (david-yz-liu) [reply .github/workflows/test.yml:41]: Also, add the `-vv` argument to this line (in addition to adding it to the next line, which is good).

File: .github/workflows/test.yml [hunk 3]
```diff
@@ -38,8 +38,8 @@ jobs:
...
+          pytest tests/test_debug -vv --ignore=tests/test_debug/test_accumulation_table.py
```
Comment (david-yz-liu) [inline .github/workflows/test.yml:42]: You don't need an `ignore` here, and instead you _should_ modify the `tests/test_debug` part so that the _only_ tests that are run are the ones in `test_accumulation_table.py`

File: CHANGELOG.md [hunk 4]
```diff
@@ -51,6 +51,8 @@ list only contains the Pylint checkers enabled by default in PythonTA.
 - Moved tests related to `__main__.py` from `test_check.py` to `test_main.py`
 - Added more unit tests to `test_main.py` to increase coverage of `__main__.py` to 100%
 - Updated `README.md` to reflect updated folder structure
+- Moved tests related to `snapshot.py` out of `test_accumulation_table.py` and into new module `test_snapshot.py`
+- Updated GitHub Action tests to ignore `test_accumulation_table.py` and add verbose output for debug testing
```
Comment (david-yz-liu) [inline CHANGELOG.md:55]: The first part ("ignore `test_accumulation_table.py`) is not the correct behaviour. Those tests should still be executed in the second `pytest` run, but ignored in the first run.

You can keep the "add verbose output" part though, that's good.

**Comment by quiz3** [comment]
I think I have now more faithfully followed the instructions. Please advise. 

I am not sure why some coverage decreased, though. Is there something I did wrong that caused that?

**Comment by david-yz-liu** [comment]
@quiz3 your changes are good. Great question about the test coverage. I looked into this: You actually need to ignore the "recursion table" tests (in a separate file under `test_debug`) in the coverage checking run of `pytest`, and instead execute those in the second run instead. I forgot that those tests were already in a separate file, but indeed the implementation of `RecursionTable` uses the same approach as `AccumulationTable` and so isn't compatible with test coverage.

**Review by david-yz-liu — APPROVED** [review]
